src/commons/authentication/oidc.rs

```diff
@@ -0,0 +1,366 @@
+use std::{
+    collections::hash_map::DefaultHasher,
+    hash::{Hash, Hasher},
+};
+
+use k8s_openapi::api::core::v1::{EnvVar, EnvVarSource, SecretKeySelector};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+use url::{ParseError, Url};
+
+#[cfg(doc)]
+use crate::commons::authentication::AuthenticationClass;
+use crate::commons::authentication::{tls::TlsClientDetails, SECRET_BASE_PATH};
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub const DEFAULT_OIDC_WELLKNOWN_PATH: &str = ".well-known/openid-configuration";
+pub const CLIENT_ID_SECRET_KEY: &str = "clientId";
+pub const CLIENT_SECRET_SECRET_KEY: &str = "clientSecret";
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("failed to parse OIDC endpoint url"))]
+    ParseOidcEndpointUrl { source: ParseError },
+
+    #[snafu(display(
+        "failed to set OIDC endpoint scheme '{scheme}' for endpoint url '{endpoint}'"
+    ))]
+    SetOidcEndpointScheme { endpoint: Url, scheme: String },
+}
+
+/// This struct contains configuration values to configure an OpenID Connect
+/// (OIDC) authentication class. Required fields are the identity provider
+/// (IdP) `hostname` and the TLS configuration. The `port` is selected
+/// automatically if not configured otherwise. The `rootPath` defaults
+/// to `/`.
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct AuthenticationProvider {
+    /// Hostname of the identity provider, e.g. `my.keycloak.corp`.
+    hostname: String,
+
+    /// Port of the identity provider. If TLS is used defaults to `443`,
+    /// otherwise to `80`.
+    port: Option<u16>,
+
+    /// Root HTTP path of the identity provider. Defaults to `/`.
+    #[serde(default = "default_root_path")]
+    root_path: String,
+
+    /// Use a TLS connection. If not specified no TLS will be used.
+    #[serde(flatten)]
+    pub tls: TlsClientDetails,
+
+    /// If a product extracts some sort of "effective user" that is represented by a
```

Comment:
Three things:

1. This also should be in the ADR
2. Yes, ideally every product (operator) should use the same identifier cluster-wide, but can be do so in the real-world. The comment mentions hard-coded claims, so we might need to think about adding it to the client-level config instead. Or am I missing something here?
3. The wording needs to change a little, but let's first resolve 1 and 2.

Reply:
Yeah, I think we are convinced that we need to support configuring the `principalClaim`. operators will error out when the product does not support the configured `principalClaim`.
Do you want to make suggestions regarding 3.), as 1.) is not under control of this PR? :)

Reply:
For reference: https://github.com/stackabletech/superset-operator/blob/fbd82b6abf8ee6b9ff9d0e54d99aa17d60cbdc5b/rust/crd/src/authentication.rs#L144-L151

Reply:
There could be a scenario where a customer already has Trino, and uses the `sub` (stable identifier) principal claim, and all rego rules that are needed. Then they want to add in Superset, which will error out if using the same AuthenticationClass.

I guess the solution would be to duplicate the AuthenticationClass, and just set the `principalClaim` accordingly.

If so, then I think we can resolve this for now, and separately discuss whether this is an AuthenticationClass level or a product level field.

Reply:
Okay, then let's stick with the current approach. This needs a followup PR in https://github.com/stackabletech/documentation to adjust the ADR.

Reply:
Created https://github.com/stackabletech/documentation/pull/498